src/vm-vcpu/src/vcpu/mod.rs

```diff
@@ -319,23 +380,63 @@ impl KvmVcpu {
                                 eprintln!("Failed to write to mmio");
                             }
                         }
-                        _ => {
+                        other => {
                             // Unhandled KVM exit.
+                            eprintln!("Unhandled vcpu exit: {:#?}", other);
                         }
                     }
                 }
                 Err(e) => {
                     // During boot KVM can exit with `EAGAIN`. In that case, do not
                     // terminate the run loop.
-                    if e == Errno::new(libc::EAGAIN) {
-                    } else {
-                        eprintln!("Emulation error: {}", e);
-                        break;
+                    match e.errno() {
+                        libc::EAGAIN => {}
+                        libc::EINTR => {
+                            interrupted_by_signal = true;
+                        }
+                        _ => {
+                            eprintln!("Emulation error: {}", e);
+                            break;
+                        }
+                    }
+                }
+            }
+
+            if interrupted_by_signal {
+                self.vcpu_fd.set_kvm_immediate_exit(0);
+                let mut run_state_lock = self.run_state.vm_state.lock().unwrap();
+                loop {
+                    match *run_state_lock {
+                        VmRunState::Running => {
+                            // The VM state is running, so we need to exit from this loop,
+                            // and enter the kvm run loop.
+                            break;
+                        }
+                        VmRunState::Suspending => {
+                            // The VM is suspending. We run this loop until we get a different
+                            // state.
+                        }
+                        VmRunState::Exiting => {
+                            // The VM is exiting. We also exit from this VCPU thread.
+                            return Ok(());
```

Comment:
We can use a `break` + `'label` here (example [here](https://doc.rust-lang.org/rust-by-example/flow_control/loop/nested.html)) to exit the outer loop. This is not super useful right now, but I think it's more idiomatic and would help if we want to do more than just return `Ok(())` before leaving the method (otherwise we'd have to duplicate that logic both here and before the `Ok(())` on line 432).